src/idfxt/JSONDataInterface.cpp

```diff
@@ -0,0 +1,209 @@
+{
+    if (attribute) {
+        string attribute_name = attribute->string;
+        if (attribute_name.find("imum")  != std::string::npos) { //only if range checkable fields
+            double attribute_value =  (attribute->valuestring) ? stod(attribute->valuestring) : attribute->valueint;
+            if (attribute_name == "exclude_minimum") {
+                if (!(property_value > attribute_value)) {
+                    valid = false;
+                    cout << "INVALID: " << child_name << " : " << property_name << " : " << property_value <<  " must be > " << attribute_value << endl;
+                }
+            } else if (attribute_name == "exclude_maximum") {
+                if (!(property_value < attribute_value)) {
+                    valid = false;
+                    cout << "INVALID: " << child_name << " : " << property_name << " : " << property_value <<  " must be < " << attribute_value << endl;
+                }
+            } else if (attribute_name == "minimum") {
+                if (!(property_value >= attribute_value)) {
+                    valid = false;
+                    cout << "INVALID: " << child_name << " : " << property_name << " : " << property_value <<  " must be >= " << attribute_value << endl;;
+                }
+            } else if (attribute_name == "maximum") {
+                if (!(property_value <= attribute_value)) {
+                    valid = false;
+                    cout << "INVALID: " << child_name << " : " << property_name << " : " << property_value <<  " must be <= " << attribute_value << endl;
+                }
+            }
+        }
+    }
+}
+
+void JSONDataInterface::checkNumeric(double property_value, string property_name, cJSON *schema_object, bool &valid, string child_name)
+{
+    if (schema_object) {                               //get matching schema property
+        cJSON *schema_property = cJSON_GetObjectItem(schema_object, property_name.c_str());
+        if (schema_property) {
+            for (int a = 0; a < cJSON_GetArraySize(schema_property); a++) {
+                cJSON *attribute = cJSON_GetArrayItem(schema_property, a);
+                checkRange(attribute, property_name, child_name, valid, property_value);
+            }
+        }
+    }
+}
+
+bool JSONDataInterface::validateModel()
+{
+    bool valid = true;
+    cJSON *child = model_j->child;
+    if (child) {
+        do {
+            string child_name = child->string;
+            for (int i = 0; i < cJSON_GetArraySize(child); i++) {
+                cJSON *property = cJSON_GetArrayItem(child,i);
+                if (property) {
+                    string property_name = property->string;
+                    switch (property->type) {
+                    case cJSON_Number: { //check numeric types
+                        double property_value = (property->valuedouble) ? property->valuedouble : property->valueint;
```

Comment:
This, again, is the result of a combination of strings, like "auto..,"
being accepted in IDF as valid values for numeric fields.  There was no
indication that this would ever be changed in E+, so when I generate the
JSON, I had to had to carry that through.  JSON itself has only a
numeric/string distinction.  This could, perhaps, be better, and further
development on all of the validation will occur on our way to September to
bring it up to par with current E+.

On Fri, Jan 16, 2015 at 6:36 PM, Dan Macumber <notifications@github.com>
wrote:

> In src/idfxt/JSONDataInterface.cpp
> <https://github.com/NREL/EnergyPlus/pull/4636#discussion-diff-23117415>:
>
> > +}
> > +
> > +bool JSONDataInterface::validateModel()
> > +{
> > +    bool valid = true;
> > +    cJSON *child = model_j->child;
> > +    if (child) {
> > +        do {
> > +            string child_name = child->string;
> > +            for (int i = 0; i < cJSON_GetArraySize(child); i++) {
> > +                cJSON *property = cJSON_GetArrayItem(child,i);
> > +                if (property) {
> > +                    string property_name = property->string;
> > +                    switch (property->type) {
> > +                    case cJSON_Number: { //check numeric types
> > +                        double property_value = (property->valuedouble) ? property->valuedouble : property->valueint;
>
> I just meant that you are converting all ints to doubles and then
> validating that way. It would be better to validate ints as ints
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NREL/EnergyPlus/pull/4636/files#r23117415>.
>